File: apps/hook_frontend/src/types/recharge.ts
export type RechargePackageStatus = 'active' | 'disabled';
export type RechargeOrderStatus = 'pending' | 'expired' | 'paid' | 'cancelled' | 'failed';
export type RechargeOrderDatePreset = 'all' | 'today' | 'last7days' | 'last30days' | 'custom';
export type PaymentCallbackStatus = 'received' | 'processed' | 'ignored' | 'failed';

export type RechargePackage = {
  id: string;
  name: string;
  description: string | null;
  recharge_amount: number;
  gift_amount: number;
  total_arrival_amount: number;
  status: RechargePackageStatus;
  sort_order: number;
  created_at: string;
  updated_at: string;
};

export type RechargeOrder = {
  id: string;
  order_no: string;
  user_id: string;
  username: string;
  user_email: string;
  package_id: string | null;
  package_name: string;
  recharge_amount: number;
  gift_amount: number;
  total_arrival_amount: number;
  payable_amount: number;
  status: RechargeOrderStatus;
  payment_channel_code: string | null;
  payment_channel_name: string | null;
  payment_method: string | null;
  provider_trade_no: string | null;
  refund_status: string | null;
  refund_amount: number | null;
  paid_at: string | null;
  refunded_at: string | null;
  expires_at: string;
  created_at: string;
  updated_at: string;
};

export type PaymentChannel = {
  code: string;
  name: string;
  enabled: boolean;
  config: Record<string, unknown>;
  secret_set: boolean;
  config_schema: PaymentChannelConfigSchema | null;
  registered_at: string;
  updated_at: string;
};

export type PublicPaymentChannel = {
  code: string;
  name: string;
  methods: PaymentMethodOption[];
};

export type PaymentCallbackRecord = {
  id: string;
  payment_channel_code: string;
  callback_kind: 'notify' | 'return';
  http_method: string;
  order_no: string | null;
  provider_trade_no: string | null;
  payment_method: string | null;
  trade_status: string | null;
  status: PaymentCallbackStatus;
  settled: boolean;
  error_message: string | null;
  raw_params: Record<string, unknown>;
  received_at: string;
  processed_at: string | null;
};

export type PaymentChannelConfigSchema = {
  fields: PaymentChannelConfigField[];
  methods: PaymentMethodOption[];
};

export type PaymentChannelConfigField = {
  key: string;
  label: string;
  secret: boolean;
  required: boolean;
};

export type PaymentMethodOption = {
  code: string;
  name: string;
};

export type RechargePackageInput = {
  name: string;
  description?: string;
  recharge_amount: number;
  gift_amount: number;
  status: RechargePackageStatus;
  sort_order: number;
};

export type PaymentChannelUpdateInput = {
  enabled: boolean;
  config?: Record<string, unknown>;
  api_key?: string;
};

export type UserRechargePackage = Omit<
  RechargePackage,
  'status' | 'created_at' | 'total_arrival_amount'
> & {
  total_arrival_amount: number;
  estimated_payable_amount: number;
};

export type RechargeOrderCreateInput = {
  package_id?: string;
  recharge_amount?: number;
  payment_channel_code: string;
  payment_method: string;
  captcha_token?: string;
};

export type PaymentOrderAction = {
  kind: 'form_post';
  action: string;
  method: 'POST';
  fields: Record<string, string>;
};

export type RechargeOrderCreateResponse = {
  order: RechargeOrder;
  payment: PaymentOrderAction;
};

export type RechargePackageListResponse = {
  items: RechargePackage[];
  total: number;
  page: number;
  page_size: number;
};

export type UserRechargePackageListResponse = {
  recharge_enabled: boolean;
  arrival_ratio: number;
  min_amount: number;
  max_amount: number;
  items: UserRechargePackage[];
  total: number;
  page: number;
  page_size: number;
};

export type RechargeOrderListResponse = {
  items: RechargeOrder[];
  total: number;
  page: number;
  page_size: number;
};

export type RechargeOrderSummary = {
  total_payable_amount: number;
  order_count: number;
  user_count: number;
};

export type RechargeOrderUserSummary = {
  user_id: string;
  username: string;
  user_email: string;
  order_count: number;
  total_payable_amount: number;
  last_paid_at: string | null;
};

export type RechargeOrderSummaryResponse = {
  summary: RechargeOrderSummary;
  items: RechargeOrderUserSummary[];
  total: number;
  page: number;
  page_size: number;
};

export type PaymentCallbackRecordListResponse = {
  items: PaymentCallbackRecord[];
  total: number;
  page: number;
  page_size: number;
};

File: crates/types/src/recharge.rs
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};

use crate::pagination::Page;

pub const RECHARGE_PACKAGE_STATUS_ACTIVE: &str = "active";
pub const RECHARGE_PACKAGE_STATUS_DISABLED: &str = "disabled";
pub const RECHARGE_ORDER_STATUS_PENDING: &str = "pending";
pub const RECHARGE_ORDER_STATUS_EXPIRED: &str = "expired";
pub const RECHARGE_ORDER_STATUS_PAID: &str = "paid";
pub const RECHARGE_ORDER_STATUS_CANCELLED: &str = "cancelled";
pub const RECHARGE_ORDER_STATUS_FAILED: &str = "failed";

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct RechargePackageListFilters {
    pub search: Option<String>,
    pub status: Option<String>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct RechargeOrderListFilters {
    pub search: Option<String>,
    pub status: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RechargePackage {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub recharge_amount: Decimal,
    pub gift_amount: Decimal,
    pub status: String,
    pub sort_order: i64,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RechargeOrder {
    pub id: String,
    pub order_no: String,
    pub user_id: String,
    pub username: String,
    pub user_email: String,
    pub package_id: Option<String>,
    pub package_name: String,
    pub recharge_amount: Decimal,
    pub gift_amount: Decimal,
    pub total_arrival_amount: Decimal,
    pub payable_amount: Decimal,
    pub status: String,
    pub payment_channel_code: Option<String>,
    pub payment_channel_name: Option<String>,
    pub expires_at: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PaymentChannel {
    pub code: String,
    pub name: String,
    pub enabled: bool,
    pub registered_at: String,
    pub updated_at: String,
}

#[derive(Clone, Debug, PartialEq, Deserialize)]
pub struct RechargePackageCreatePayload {
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(with = "rust_decimal::serde::float")]
    pub recharge_amount: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub gift_amount: Decimal,
    #[serde(default)]
    pub status: Option<String>,
    pub sort_order: i64,
}

#[derive(Clone, Debug, PartialEq, Deserialize)]
pub struct RechargePackageUpdatePayload {
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(with = "rust_decimal::serde::float")]
    pub recharge_amount: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub gift_amount: Decimal,
    pub status: String,
    pub sort_order: i64,
}

#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
pub struct PaymentChannelUpdatePayload {
    pub enabled: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
pub struct RechargeOrderCreatePayload {
    pub package_id: String,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct RechargePackageResponse {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    #[serde(with = "rust_decimal::serde::float")]
    pub recharge_amount: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub gift_amount: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub total_arrival_amount: Decimal,
    pub status: String,
    pub sort_order: i64,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct RechargeOrderResponse {
    pub id: String,
    pub order_no: String,
    pub user_id: String,
    pub username: String,
    pub user_email: String,
    pub package_id: Option<String>,
    pub package_name: String,
    #[serde(with = "rust_decimal::serde::float")]
    pub recharge_amount: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub gift_amount: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub total_arrival_amount: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub payable_amount: Decimal,
    pub status: String,
    pub payment_channel_code: Option<String>,
    pub payment_channel_name: Option<String>,
    pub expires_at: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub struct PaymentChannelResponse {
    pub code: String,
    pub name: String,
    pub enabled: bool,
    pub registered_at: String,
    pub updated_at: String,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct RechargePackageListResponse {
    pub items: Vec<RechargePackageResponse>,
    pub total: u64,
    pub page: u64,
    pub page_size: u64,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct UserRechargePackageResponse {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    #[serde(with = "rust_decimal::serde::float")]
    pub recharge_amount: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub gift_amount: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub total_arrival_amount: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub estimated_payable_amount: Decimal,
    pub sort_order: i64,
    pub updated_at: String,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct UserRechargePackageListResponse {
    pub recharge_enabled: bool,
    #[serde(with = "rust_decimal::serde::float")]
    pub arrival_ratio: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub min_amount: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub max_amount: Decimal,
    pub items: Vec<UserRechargePackageResponse>,
    pub total: u64,
    pub page: u64,
    pub page_size: u64,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct RechargeOrderListResponse {
    pub items: Vec<RechargeOrderResponse>,
    pub total: u64,
    pub page: u64,
    pub page_size: u64,
}

impl From<RechargePackage> for RechargePackageResponse {
    fn from(value: RechargePackage) -> Self {
        let total_arrival_amount = value.recharge_amount + value.gift_amount;
        Self {
            id: value.id,
            name: value.name,
            description: value.description,
            recharge_amount: value.recharge_amount,
            gift_amount: value.gift_amount,
            total_arrival_amount,
            status: value.status,
            sort_order: value.sort_order,
            created_at: value.created_at,
            updated_at: value.updated_at,
        }
    }
}

impl From<RechargeOrder> for RechargeOrderResponse {
    fn from(value: RechargeOrder) -> Self {
        Self {
            id: value.id,
            order_no: value.order_no,
            user_id: value.user_id,
            username: value.username,
            user_email: value.user_email,
            package_id: value.package_id,
            package_name: value.package_name,
            recharge_amount: value.recharge_amount,
            gift_amount: value.gift_amount,
            total_arrival_amount: value.total_arrival_amount,
            payable_amount: value.payable_amount,
            status: value.status,
            payment_channel_code: value.payment_channel_code,
            payment_channel_name: value.payment_channel_name,
            expires_at: value.expires_at,
            created_at: value.created_at,
            updated_at: value.updated_at,
        }
    }
}

impl From<PaymentChannel> for PaymentChannelResponse {
    fn from(value: PaymentChannel) -> Self {
        Self {
            code: value.code,
            name: value.name,
            enabled: value.enabled,
            registered_at: value.registered_at,
            updated_at: value.updated_at,
        }
    }
}

impl From<Page<RechargePackage>> for RechargePackageListResponse {
    fn from(value: Page<RechargePackage>) -> Self {
        Self {
            items: value.items.into_iter().map(Into::into).collect(),
            total: value.total,
            page: value.page,
            page_size: value.page_size,
        }
    }
}

impl From<Page<RechargeOrder>> for RechargeOrderListResponse {
    fn from(value: Page<RechargeOrder>) -> Self {
        Self {
            items: value.items.into_iter().map(Into::into).collect(),
            total: value.total,
            page: value.page,
            page_size: value.page_size,
        }
    }
}
